wagtail/models/sites.py

```diff
@@ -213,54 +261,64 @@ def get_site_root_paths():
-                    result.append(
+        cached_result = cls._get_cached_site_root_paths()
+        if cached_result is not None:
+            return deepcopy(cached_result)
+
+        result = []
+        for site in Site.objects.get_cached_list():
+            if getattr(settings, "WAGTAIL_I18N_ENABLED", False):
+                result.extend(
+                    [
                         SiteRootPath(
                             site.id,
-                            site.root_page.url_path,
+                            root_page.url_path,
                             site.root_url,
-                            site.root_page.locale.language_code,
+                            root_page.locale.language_code,
                         )
+                        for root_page in site.root_page.get_translations(
+                            inclusive=True
+                        ).select_related("locale")
+                    ]
+                )
+            else:
+                result.append(
+                    SiteRootPath(
+                        site.id,
+                        site.root_page.url_path,
+                        site.root_url,
+                        site.root_page.locale.language_code,
                     )
+                )
 
-            cache.set(
-                SITE_ROOT_PATHS_CACHE_KEY,
-                result,
-                3600,
-                version=SITE_ROOT_PATHS_CACHE_VERSION,
-            )
-
-        else:
-            # Convert the cache result to a list of SiteRootPath tuples, as some
-            # cache backends (e.g. Redis) don't support named tuples.
-            result = [SiteRootPath(*result) for result in result]
+        # cache result for the next request
+        _site_root_paths_cache.value = result
 
         return result
 
     @staticmethod
-    def clear_site_root_paths_cache():
-        cache.delete(SITE_ROOT_PATHS_CACHE_KEY, version=SITE_ROOT_PATHS_CACHE_VERSION)
+    def _get_cached_site_root_paths() -> Union[List[SiteRootPath], None]:
+        return getattr(_site_root_paths_cache, "value", None)
+
+    @staticmethod
+    def clear_site_root_paths_cache() -> None:
+        """
+        As a courtesy, we clear cached `SiteRootPath` data for threads currently
+        being handled by this instance when changes that might affect the return
+        value are detected.
+
+        Any threads running in other instances that already have cached result
+        will continue to use that data for the remaining life of the thread. But,
+        any new requests/threads (or active threads that have not year queried
+        the cache) should pick up the updated values.
+        """
+        _site_root_paths_cache.clear()
+
+    @staticmethod
+    def clear_caches():
+        """
+        Convenience method for clearing both the `Site` and `SiteRootPath`
+        threadlocal caches.
+        """
+        _sites_cache.clear()
+        _site_root_paths_cache.clear()
```

Comment:
Suggestion: Given we have additional methods for clearing the specific caches, I suggest we reuse those

Reply:
I must admit I'm on the fence with this one. I like the clarity here... Calling a manager method from the class feels a little awkward.

Reply:
I'm gonna keep this bit as it is unless anyone feels strongly about it?